field/TinyLang/Field/Evaluator.hs

```diff
@@ -204,72 +265,97 @@ denoteUniConst (UniConst Vector v) = unAField . fromInteger . packPositiveAsc $
 denoteSomeUniConst :: Field f => SomeUniConst f -> f
 denoteSomeUniConst = forget denoteUniConst
 
-denoteExpr
-    :: (MonadEvalError f m, Eq f, Field f, AsInteger f)
-    => Env (SomeUniConst f) -> Expr f a -> m f
-denoteExpr env = fmap denoteUniConst . evalExprUni env
-
-normStatement
-    :: (MonadEvalError f m, Eq f, Field f, AsInteger f)
-    => Env (SomeUniConst f) -> Statement f -> Expr f a -> m (Expr f a)
-normStatement env (ELet (UniVar uni var) def) rest = do
-    defN <- normExpr env def
+denoteExpr :: (Monad m, Eq f, Field f, AsInteger f)
+    => Expr f a -> EvalT f m f
+denoteExpr = fmap denoteUniConst . evalExprUni
+
+-- | A recursive normalizer for programs.
+normProgram :: (Monad m, Eq f, Field f, AsInteger f)
+    => Program f -> EvalT f m (Program f)
+normProgram (Program stmts) = normStatements stmts (pure . Program)
+
+-- | A recursive normalizer for statements
+normStatements :: (Monad m, Eq f, Field f, AsInteger f)
+    => Statements f -> (Statements f -> EvalT f m r) -> EvalT f m r
+normStatements (Statements []) kont       = kont $ Statements []
+normStatements (Statements (s:rest)) kont = normStatement s $ \case
+          Nothing -> normStatements (Statements rest) kont
+          Just sN -> normStatements (Statements rest) $ \ (Statements restN) ->
+              kont $ Statements $ sN : restN
```

Comment:
You could avoid all this wrapping-unwrapping by using the worker-wrapper transformation and having a local function like

```haskell
go :: [Statement f] -> ([Statement f] -> EvalT f m r) -> EvalT f m r
```